schemas/styles/csl.rnc

```diff
@@ -649,6 +650,62 @@ div {
 }
 # ==============================================================================
 
+## cs:title Rendering Element
+div {
+  
+  ## Use to render titles if "title-main" and "title-sub" need different formatting
+  ## or to omit either "title-main" or "title-sub"
+  rendering-element.title =
+    element cs:title {
+      title.attributes, title.formatting, (title.main?, title.sub?)
+    }
+  title.attributes =
```

Comment:
So these are the attributes associated with the proposed new `cs:title` element. 

And they are basically instructions for how processors should parse titles strings into structured data.

Right?

I think for all of us, there are two questions:

1. do we really want to do this? why can't we just demand that processors get properly structured data? 
2. if yes, do we need a dedicated element to hold these attributes? For sake of argument, could they not go on `style`, `citation`, and/or `bibliography`? Related to my broader question about elements vs attributes.

Reply:
I started to try to draft a git commit message that would concisely explain the changes, but got a bit stuck. We can figure out how to resolve once others weigh in.

> Add support for configuration of complex titles
>                                               
> This adds ability to configure title and subtitle rendering independently. 

> It does this by adding 'title-main' and 'title-sub' and relates variables.                                                              
>                                                                           
>  It also adds new elements and attributes that specify ...

Reply:
> So these are the attributes associated with the proposed new `cs:title` element.
> 
> ~And they are basically instructions for how processors should parse titles strings into structured data.~
> 
> ~Right?~ Well, maybe a bit more than that.

Yes, they instruct processors where they should split titles into segments.

> I think for all of us, there are two questions:
> 
> 1. do we really want to do this? why can't we just demand that processors get properly structured data?

Because that would involve talking to and convincing a lot of people (mainly Zotero, Mendeley, and Papers, but basically everyone on the list on citationstyles.org). They would have to change their internal data structure and their UI fields, or do the parsing themselves before they pass the data to the processor. 
What is more, different styles have different opinions about what exactly constitutes a title or where a subtitle begins. So if we really want style agnostic data, then doing this in the processor is the only real option.
Next, there is legacy data and all of this uses titles in an unstructured single field. With these changes, users will just be able to use the same data as before in new ways. (E.g. A scholar in the social sciences who has added titles according to APA standards, i.e. with the original punctuation, can now produce correct references with Chicago author date just by selecting the correct style.)

> 2. if yes, do we need a dedicated element to hold these attributes? For sake of argument, could they not go on `style`, `citation`, and/or `bibliography`? Related to my broader question about elements vs attributes.

Well, we do that as well. If you look at line 956, you'll find a couple of inheritable options that so essentially the same thing. The idea is that you'll use those on `style`, `citation`, `bibliography` if you render titles with `cs:text`. That should be what do in most cases anyway. `cs:title` should really only be needed if you have special requirements, like different formatting for title vs. subtitle.

Reply:
Maybe an idea: The main use case for `cs:title` is the ABNT style. I don't know other details about this style, but your suggestion from above could work, if a couple of requirements are met:
- they use exclusively use one specific character between title and subtitle, e.g. a colon.
- they normalize all the characters they recognize as splitting characters

Also: I don't know about other styles that have similar requirements.

If we do this, we will also have to make `title-main` and `title-sub` available in styles, which they currently are not.

Reply:
@bwiernik What do you think about this here?

Reply:
The suggestion being don't have a `cs:title` rendering element and instead expose `-main` and `-sub` as first class variables? How does that handle Chicago normalizing punctuation?

Reply:
I don't think it would affect that at all. You can still use `<text variable="title">` to render the complete title with punctuation normalized. You just can't have independent formatting of title and subtitle.

Reply:
So, for ABNT, it would be:
```
<group delimiter=": ">
  <text variable='title-main" font-weight="bold"/>
  <text variable='title-sub"/>
</group>
```

Would any delimiters in `title-sub` be normalized?

Reply:
Depending on the attributes, yes, I think they should be normalized.

Reply:
> `cs:title` should really only be needed if you have special requirements, like different formatting for title vs. subtitle.

I'm not following here. You mean a specific kind of "different formatting"; don't you?

Reply:
So what can't it do?

Reply:
@bdarcus title bold, subtitle normal font

@bwiernik certain combinations of requirements won't work. APA style not normalization of punctuation and ABNT style title formatting (title bold, subtitle normal).

Reply:
The only people who've ever asked for separate styling, as you say, is ABNT. I think for now we can reasonably go without `cs:title`. If it ever comes up again, we can revisit `cs:title`.

Reply:
One thing I'd like to verify. Is it okay for the `title-delimiter` and `title-sub-delimiters` to be empty by default on a style? Are empty attributes acceptable? @bdarcus

Reply:
> Is it okay for the `title-delimiter` and `title-sub-delimiters` to be empty by default on a style? Are empty attributes acceptable?

What exactly do you mean @bwiernik? 

Acceptable to what or whom?

And what do you mean by "by default"?

Certainly this is valid XML;

``` xml
<foo bar=""/>
```

Reply:
Isn't that something citeproc have to take care if? Like "if the attribute is there then ... otherwise ..."?

Reply:
Currently, the proposal is, if `title-delimiter` is not assigned on the style (that is, not present at all), then no normalization occurs. I guess that is valid, because styles can have `default-locale` absent.

Reply:
> @bdarcus title bold, subtitle normal font

I'm admittedly pretty tired today, so might just be me, but why wouldn't this allow just that?

``` xml
<group delimiter=": ">
  <text variable='title-main" font-weight="bold"/>
  <text variable='title-sub"/>
</group>
```

Reply:
It does, but with certain limitations as outlined above. But as it turns out these limitations are of only theoretical relevance and won't really matter much. So, yeah, we should go that route.